src/Microsoft.OData.Core/JsonLight/ODataJsonLightReader.cs

```diff
@@ -2426,20 +2426,18 @@ private void StartNestedStreamInfo(ODataJsonLightReaderStreamInfo readerStreamIn
         /// <returns>True if successfully append key segment.</returns>
         private bool TryAppendEntitySetKeySegment(ref ODataPath odataPath)
         {
-            try
+            if (EdmExtensionMethods.HasKey(this.CurrentScope.NavigationSource, this.CurrentScope.ResourceType as IEdmStructuredType))
             {
-                if (EdmExtensionMethods.HasKey(this.CurrentScope.NavigationSource, this.CurrentScope.ResourceType as IEdmStructuredType))
+                IEdmEntityType currentEntityType = this.CurrentScope.ResourceType as IEdmEntityType;
+                ODataResourceBase resource = this.CurrentScope.Item as ODataResourceBase;
+                KeyValuePair<string, object>[] keys = ODataResourceMetadataContext.GetKeyProperties(resource, null, currentEntityType, false);

```

Comment:
`GetKeyProperties` seems to make the assumption that `resource` isn't `null`. I know that this isn't really related to your change, but not validating that `currentEntityType` is actually a `IEdmEntityType` means we could get a `NullReferenceException` here.

Reply:
This is our code and should never be called with a null resource. I added an assert to GetKeyProperties.